### PR TITLE
refactor: migrate markers as first class plugins

### PR DESCRIPTION
**Pull request by miurahr**
- Deprecate Core.registerMarker and Core.getMarkers
- Remove MarkerController.init and WhitespaceMarkerFactory.init

## Pull request type

- Other (describe below)
refactor

## What does this PR change?

- Make `NBSPMarker` as true plugin
- NBSP marker menu are created in the plugin
- Removal of hard coded menu creation
- When registering maker plugins with its class, and loading the plugin, static fields are initialized when `loadPlugins` static function is called. it breaks colors of markers.

### CODE REVIEW

File: src/org/omegat/gui/editor/MarkerController.java [hunk 1]
```diff
@@ -116,6 +135,26 @@ public static void init() throws Exception {
...
         }
+
+        // default markers
+        JCheckBoxMenuItem menuItem = new JCheckBoxMenuItem();
+        nbspMarker.setEnabled(Preferences.isPreference(Preferences.MARK_NBSP));
+        menuItem.setSelected(nbspMarker.isEnabled());
+        menuItem.setIcon(MainMenuIcons.newColorIcon(Styles.EditorColor.COLOR_NBSP.getColor()));
+        Mnemonics.setLocalizedText(menuItem, OStrings.getString("MW_VIEW_MENU_MARK_NBSP"));
+        menuItem.addActionListener(e -> {
+            isNBSP = menuItem.isSelected();
```
Comment (miurahr) [inline src/org/omegat/gui/editor/MarkerController.java:146]: It is necessary to improve more.

**Comment by miurahr** [comment]
Now these 6 maker become true plugins.

```
    org.omegat.core.spellchecker.SpellCheckerMarker \
    org.omegat.gui.glossary.TransTipsMarker \
    org.omegat.gui.editor.mark.BidiMarkers \
    org.omegat.gui.editor.mark.ComesFromAutoTMMarker \
    org.omegat.gui.editor.mark.FontFallbackMarker \
    org.omegat.gui.editor.mark.ReplaceMarker \
```

Menu Items are generated dynamically using  marker plugins IMarker interface.

**Comment by miurahr** [comment]
I propose addition of method `    void reloadDocument();` in `IEditor` core interface, that reload project contents and redraw markers when project is loaded.

**Comment by brandelune** [comment]
@t-cordonnier review appreciated ! :-)

**Comment by t-cordonnier** [comment]
> @t-cordonnier review appreciated ! :-)

@brandelune  I already sent some comments in this ticket. 

I notice that now the menus are more dynamic as I suggested, good point. So, there was an evolution since my last comments, which I was not alerted about, so for me it was still pending.

The rest of the code seems unchanged so I maintain what I said: I would prefer one or more sub-interfaces to distinguish between non-skippable markers (the ones which are always active), menu markers (the ones which have a menu with an icon to activate/inactivate) and eventually, those which are skippable but whose configuration is done in another location than in a menu.
I remember in the past Aaron wanted to avoid casts, but maybe @miurahr can have a different position. I let to him the final decision but for the moment I saw not even a reaction to my comments.

**Comment by miurahr** [comment]
The change does not provide a benefit over the task.

**Pull request closed without merging**